worktree_commit.go

```diff
@@ -0,0 +1,243 @@
+package git
+
+import (
+	"io"
+	"path/filepath"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+	"gopkg.in/src-d/go-git.v4/plumbing/filemode"
+	"gopkg.in/src-d/go-git.v4/plumbing/format/index"
+	"gopkg.in/src-d/go-git.v4/plumbing/object"
+	"gopkg.in/src-d/go-git.v4/storage"
+	"gopkg.in/src-d/go-git.v4/utils/ioutil"
+
+	"gopkg.in/src-d/go-billy.v2"
+)
+
+// Commit stores the current contents of the index in a new commit along with
+// a log message from the user describing the changes.
+func (w *Worktree) Commit(msg string, opts *CommitOptions) (plumbing.Hash, error) {
```

Comment:
You can have that if you want, with wrapper functions.  But you don't actually need to write the core logic like that to get that usage pattern.  The whole purpose of the index is to isolate the worktree from the repository:

 - there is a set of conceptual operations between the worktree and the index (like git add).

 - there is a different set of conceptual operations between the index and the repository (like git commit).

(except with a few concessions here and there due to pure usability that are always implemented at a very high level)

See, passing the index to the commit function avoids relaying on internal state and make it much easier to test and to mock.  Then moving the commit function to a class that is not the worktree will be help to distribute responsibilities where they originally belong in the real git.